/*
 * dedupv1 - iSCSI based Deduplication System for Linux
 *
 * (C) 2008 Dirk Meister
 * (C) 2009 - 2011, Dirk Meister, Paderborn Center for Parallel Computing
 */

 // compile with /opt/dedupv1/bin/protoc --cpp_out=. dedupv1.proto with working directory common/resources

message DiskHashIndexLogfileData {
    // for configuration checking
    optional uint32 page_size = 1;
    optional uint64 size = 2;
    repeated string filename = 3;
    optional bool overflow_area = 4;
}

message DiskHashPageData {
    optional fixed32 crc = 1;
    optional uint32 entry_count = 2;
    optional bool overflow = 3 [default = false];
}

message DiskHashTransactionPageData {
    optional uint64 bucket_id = 1;
    optional fixed32 original_crc = 2;
    optional fixed32 transaction_crc = 3;
    optional bytes data = 4;
    optional uint64 item_count = 5;
    optional uint64 version = 6;
}

message FixedIndexMetaData {
    optional uint64 width = 1;
    optional uint64 size = 2;
    optional uint64 file_count = 3;
}

enum FixedIndexBucketStateData {
   FIXED_INDEX_STATE_VALID = 0;
   FIXED_INDEX_STATE_INVALID = 1;
}

message FixedIndexBucketData {
    optional int64 key = 1;
    optional bytes data = 2;
    optional FixedIndexBucketStateData state = 3;
    optional fixed32 crc = 4;
}

message IntData {
    optional uint64 i = 1;
}

message BitmapData {
    optional uint64 size = 1;
    optional uint64 clean_bits = 2;
    optional uint64 page_size = 3;
    optional uint32 crc = 4;
}

message BitmapPageData {
    optional bytes data = 1;
    optional uint32 crc = 2;
}

// used only for testing
message RawByteData {
    optional string raw_value = 1;
}
